internal/translation/datafederation/conversion.go

```diff
@@ -0,0 +1,371 @@
+package datafederation
+
+import (
+	"fmt"
+	"reflect"
+
+	"go.mongodb.org/atlas-sdk/v20231115008/admin"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/cmp"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/pointer"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1/common"
+)
+
+type DataFederation struct {
+	*akov2.DataFederationSpec
+	ProjectID string
+	Hostnames []string
+}
+
+// SpecEqualsTo returns true if the spec of the data federation instance semantically equals to the given one.
+// Note: it assumes the spec is already normalized.
+func (df *DataFederation) SpecEqualsTo(target *DataFederation) bool {
+	dfCopy, toCopy := df.DeepCopy(), target.DeepCopy()
+	return reflect.DeepEqual(pruneSpec(dfCopy), pruneSpec(toCopy))
```

Comment:
In a perfect world we would do this. But the internal translation layer type is a composite of the CRD structs so we have to remove some fields to make it semantically comparable. Normalization only brings "order" inside the struct, most notably ordering all nested slices in a deterministic manner. If we would strip this data, we wouldn't be able to do a roundtrip conversion back to a CRD because we would lose data.